### PR TITLE
[MM-56348] system/ping: add new method with options

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
Added a new client method with options for the `GetSystemPing` so that we can stop accumulating many new methods if we decide to extend it :)

Apart from that added a new query parameter for the `/ping` endpoint to conform REST semantics.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-56348

#### Release Note

```release-note
mmctl system ping will return detailed server status even if the server status is unhealthy.
```

### CODE REVIEW

File: server/channels/api4/system.go [hunk 1]
```diff
@@ -194,7 +194,7 @@ func getSystemPing(c *Context, w http.ResponseWriter, r *http.Request) {
```
Comment (agnivade) [inline server/channels/api4/system.go:194]: I can't comment above, but I think the issue is in L182 where we don't actually log the `appErr`. So there is no indication of exactly what went wrong.

**Review by agnivade — CHANGES_REQUESTED** [review]

File: server/channels/api4/system.go (hunk 1 above)
Comment (isacikgoz) [reply server/channels/api4/system.go:194]: Right.. I totally forget that part :)

File: api/v4/source/system.yaml [hunk 2]
```diff
@@ -47,6 +47,11 @@
 
         __Minimum server version__: 6.5
 
+        If "keep_rest_semantics" is set to true in the query, the endpoint will not return
```
Comment (hanzei) [inline api/v4/source/system.yaml:50]: `keep_rest_semantics` seems a bit misleading to me, as the name indicates that the semantics have changed recently. What about `use_rest_semantics`?

File: server/public/model/client4.go [hunk 3]
```diff
@@ -4608,37 +4608,42 @@ func (c *Client4) GenerateSupportPacket(ctx context.Context) ([]byte, *Response,
...
 
 // GetPingWithFullServerStatus will return the full status if several basic server
 // health checks all pass successfully.
 func (c *Client4) GetPingWithFullServerStatus(ctx context.Context) (map[string]string, *Response, error) {
-	r, err := c.DoAPIGet(ctx, c.systemRoute()+"/ping?get_server_status="+c.boolString(true), "")
+	return c.GetPingWithOptions(ctx, SystemPingOptions{FullStatus: true})
+}
+
+// GetPingWithOptions will return the status according to the options
+func (c *Client4) GetPingWithOptions(ctx context.Context, options SystemPingOptions) (map[string]string, *Response, error) {
```
Comment (hanzei) [inline server/public/model/client4.go:4637]: We got four methods, all hitting the same endpoint. Maybe it's worth simplifying to only `GetPingWithOptions`.

**Review by hanzei — CHANGES_REQUESTED** [review]
Nice work, Ibrahim :+1: 

It's a bummer that the semantics of the endpoint are not ready REST-like. Your solution with a flag is a good way to avoid a breaking change :+1:

File: api/v4/source/system.yaml (hunk 2 above)
Comment (isacikgoz) [reply api/v4/source/system.yaml:50]: Makes sense.

File: server/public/model/client4.go (hunk 3 above)
Comment (isacikgoz) [reply server/public/model/client4.go:4637]: Yes, it's to keep backwards compatibility, the other methods already calling the new `GetPingWithOptions` method. In time we will probably delete others.

Comment (hanzei) [reply server/public/model/client4.go:4637]: 1/5 that we then should add a deprecation warning to the methods that we want to remove.

File: api/v4/source/system.yaml (hunk 2 above)
Comment (hanzei) [inline api/v4/source/system.yaml:50]:
```suggestion
        If "use_rest_semantics" is set to true in the query, the endpoint will not return
```

File: api/v4/source/system.yaml [hunk 4]
```diff
@@ -64,6 +69,12 @@
           required: false
           schema:
             type: string
+        - name: keep_rest_semantics
```
Comment (hanzei) [inline api/v4/source/system.yaml:72]:
```suggestion
        - name: use_rest_semantics
```

**Review by hanzei — CHANGES_REQUESTED** [review]
One thing that I forgot: We should update the webapp to use `use_rest_semantics`

https://github.com/mattermost/mattermost/blob/845f28cb7617b7d7d06d9d7aef0c1da6114f6d38/webapp/platform/client/src/client4.ts#L2452-L2462

Then we can also remove the `ignoreStatus` option in the webapp.

File: server/public/model/client4.go (hunk 3 above)
Comment (isacikgoz) [reply server/public/model/client4.go:4640]: Right, added 👍

**Comment by isacikgoz** [comment]
> One thing that I forgot: We should update the webapp to use `use_rest_semantics`

@hanzei since this is not a breaking change we don't need to update the webapp code. Am I missing something?

**Comment by hanzei** [comment]
The webapp has some special handling when calling the `/ping` endpoint (via `ignoreStatus`) as it ignores the returned status (which sometimes may not 200). My request is to remove the `ignoreStatus` from `doFetch` and therefore remove the special handling.

**Comment by isacikgoz** [comment]
cc @saturninoabril There is one test seem to failed, seems like not related to my changes but wanted to ping you if we should look deeper into it.

**Comment by saturninoabril** [comment]
Thanks for the ping. Yes not related. It's a known issue and something we'll clean up soon as part of E2E GHA migration.

**Pull request merged**